src/algebra/ring_quot.lean

```diff
@@ -0,0 +1,285 @@
+agrees with the quotient by the appropriate ideal.
+-/
+
+variables {B : Type u₁} [comm_ring B]
+
+/-- The universal ring homomorphism from `ring_quot r` to `(ideal.of_rel r).quotient`. -/
+def ring_quot_to_ideal_quotient (r : B → B → Prop) :
+  ring_quot r →+* (ideal.of_rel r).quotient :=
+lift
+  (ideal.quotient.mk (ideal.of_rel r))
+  (λ x y h, quot.sound (submodule.mem_Inf.mpr (λ p w, w ⟨x, y, h, rfl⟩)))
+
+@[simp] lemma ring_quot_to_ideal_quotient_apply (r : B → B → Prop) (x : B) :
+  ring_quot_to_ideal_quotient r (mk_ring_hom r x) = ideal.quotient.mk _ x := rfl
+
+/-- The universal ring homomorphism from `(ideal.of_rel r).quotient` to `ring_quot r`. -/
+def ideal_quotient_to_ring_quot (r : B → B → Prop) :
+  (ideal.of_rel r).quotient →+* ring_quot r :=
+ideal.quotient.lift (ideal.of_rel r) (mk_ring_hom r)
+begin
+  intros x h,
+  apply submodule.span_induction h,
+  { rintros - ⟨a,b,h,rfl⟩,
+    rw [ring_hom.map_sub, mk_ring_hom_rel h, sub_self], },
+  { simp, },
+  { intros a b ha hb, simp [ha, hb], },
+  { intros a x hx, simp [hx], },
+end
+
+@[simp] lemma ideal_quotient_to_ring_quot_apply (r : B → B → Prop) (x : B) :
+  ideal_quotient_to_ring_quot r (ideal.quotient.mk _ x) = mk_ring_hom r x := rfl
+
+/--
+The ring equivalence between `ring_quot r` and `(ideal.of_rel r).quotient`
+-/
+def ring_quot_equiv_ideal_quotient (r : B → B → Prop) :
+  ring_quot r ≃+* (ideal.of_rel r).quotient :=
+ring_equiv.of_hom_inv (ring_quot_to_ideal_quotient r) (ideal_quotient_to_ring_quot r)
+  (by { ext, simp, }) (by { ext ⟨x⟩, simp, })
+
+end comm_ring
+
+section algebra
+
+variables (S)
+
+/--
+The quotient map from an `S`-algebra to its quotient, as a homomorphism of `S`-algebras.
+-/
+def mk_alg_hom (s : A → A → Prop) : A →ₐ[S] ring_quot s :=
+{ commutes' := λ r, rfl,
+  ..mk_ring_hom s }
+
+@[simp]
+lemma mk_alg_hom_coe (s : A → A → Prop) : (mk_alg_hom S s : A →+* ring_quot s) = mk_ring_hom s :=
+rfl
+
+lemma mk_alg_hom_rel {s : A → A → Prop} {x y : A} (w : s x y) :
+  mk_alg_hom S s x = mk_alg_hom S s y :=
+quot.sound (rel.of w)
+
+lemma mk_alg_hom_surjective (s : A → A → Prop) : function.surjective (mk_alg_hom S s) :=
+by { dsimp [mk_alg_hom], rintro ⟨a⟩, use a, refl, }
+
+variables {B : Type u₄} [semiring B] [algebra S B]
+
+@[ext]
+lemma ring_quot_ext' {s : A → A → Prop}
+  (f g : ring_quot s →ₐ[S] B) (w : f.comp (mk_alg_hom S s) = g.comp (mk_alg_hom S s)) : f = g :=
+begin
+  ext,
+  rcases mk_alg_hom_surjective S s x with ⟨x, rfl⟩,
+  exact (congr_arg (λ h : A →ₐ[S] B, h x) w), -- TODO should we have `alg_hom.congr` for this?
+end
+
+/--
+Any `S`-algebra homomorphism `f : A →ₐ[S] B` which respects a relation `s : A → A → Prop`
+factors through a morphism `ring_quot s →ₐ[S]  B`.
+-/
+def lift_alg_hom (f : A →ₐ[S] B) {s : A → A → Prop} (w : ∀ ⦃x y⦄, s x y → f x = f y) :
+  ring_quot s →ₐ[S] B :=
+{ to_fun := quot.lift f
+  begin
+    rintros _ _ r,
+    induction r with a b r a b c r r' a b c r r' a b c r r',
```

Comment:
Can you use the same trick as line 140-143 here?

Reply:
I think that makes sense. After that: LGTM

bors d+